packages/block-library/src/gallery/tiles.native.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * External dependencies
+ */
+import { View, StyleSheet } from 'react-native';
+
+/**
+ * WordPress dependencies
+ */
+import { Children } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import styles from './tiles-styles';
+
+function Tiles( props ) {
+	const {
+		columns,
+		children,
+		spacing = 10,
+		style,
+	} = props;
+
+	const { compose } = StyleSheet;
+
+	const tileCount = Children.count( children );
+	const lastTile = tileCount - 1;
+	const lastRow = Math.floor( lastTile / columns );
+
+	const wrappedChildren = Children.map( children, ( child, index ) => {
+		/** Since we don't have `calc()`, we must calculate our spacings here in
+		 * order to preserve even spacing between tiles and equal width for tiles
+		 * in a given row.
+		 *
+		 * In order to ensure equal sizing of tile contents, we distribute the
+		 * spacing such that each tile has an equal "share" of the fixed spacing.
+		 * To keep the tiles properly aligned within their rows, we calculate the
+		 * left and right border widths based on the tile's relative position within
+		 * the row.
+		 *
+		 * Note: we use transparent borders instead of margins so that the fixed
+		 * spacing is included within the relative spacing (i.e. width percentage),
+		 * and wrapping behavior is preserved.
+		 *
+		 *  - The left most tile in a row must have left border width of zero.
+		 *  - The right most tile in a row must have a right border width of zero.
+		 *
+		 * The values of these widths are interpolated for tiles in between. The
+		 * right border width is complementary with the left border width of the
+		 * next tile (i.e. the right border of [tile n] + the left border of
+		 * [tile n + 1] will be equal for all tiles except the last one in a given
+		 * row).
+		 */
+
+		const row = Math.floor( index / columns );
+		const rowLength = row === lastRow ? ( lastTile % columns ) + 1 : columns;
+		const indexInRow = index % columns;
+
+		return (
+			<View style={ [ styles.tileStyle, {
```

Comment:
We should set a key here otherwise we are getting a severe warning:

<img width="727" alt="Screen Shot 2019-11-25 at 19 44 22" src="https://user-images.githubusercontent.com/5032900/69559984-0592da80-0fbc-11ea-9d8a-a42d95282c5c.png">

Adding `key={ index }` should solve it.

Reply:
Thanks for documenting that. Hm.. :thinking: that's unexpected.

> Adding key={ index } should solve it.

The weird thing is that this should be the default. From [the docs](https://reactjs.org/docs/lists-and-keys.html#keys):

> If you choose not to assign an explicit key to list items then React will default to using indexes as keys.

I haven't been able to reproduce it. Can you provide steps to reproduce it?

Reply:
I am trying with Android emulator Pixel 2XL and enabled Remote JS Debugging.

I created a Gallery with 7 images.

Whenever I open the post or rotate the device that error is printed in Chrome dev tools console.

Thanks for the docs btw. We can use web's approach to set keys, it is a lot better as far as I see: `key={ img.id || img.url }` (from .../src/gallery/gallery.js).

Reply:
I tried with those steps, but was still unable to reproduce this.

> We can use web's approach to set keys, it is a lot better as far as I see: key={ img.id || img.url }

We are currently using [this logic as well in gallery.native.js](https://github.com/WordPress/gutenberg/pull/18176/files#diff-c224ac948fbbffd7b92bbb5e04763319R66) (though there may be [some concerns about it](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1610#issuecomment-558532685)). Since we are currently wrapping the children in a `View`, I'd expect the children to be assigned their default key as index. Do you have the html content of the post where you see this warning? Perhaps the content can provide a clue about how this is occurring, or make it possible to reproduce.

It's not clear why React is not using the default `index` as key, but another possibility is to elevate the key to the child's wrapping view:

```diff
diff --git a/packages/block-library/src/gallery/tiles.native.js b/packages/block-library/src/gallery/tiles.native.js
index f1666db..895f126 100644
--- a/packages/block-library/src/gallery/tiles.native.js
+++ b/packages/block-library/src/gallery/tiles.native.js
@@ -56,7 +56,7 @@ function Tiles( props ) {
 		const indexInRow = index % columns;
 
 		return (
-			<View style={ [ styles.tileStyle, {
+			<View key={ child.key } style={ [ styles.tileStyle, {
 				width: `${ 100 / rowLength }%`,
 				paddingLeft: spacing * ( indexInRow / rowLength ),
 				paddingRight: spacing * ( 1 - ( ( indexInRow + 1 ) / rowLength ) ),
```

But, since I can't reproduce the original warning, I couldn't confirm whether this addresses the issue you've observed.

Reply:
```
<!-- wp:gallery {"ids":["708","415","701","728","684","682","416",555,"737"],"columns":3,"linkTo":"media"} -->
<figure class="wp-block-gallery columns-3 is-cropped"><ul class="blocks-gallery-grid"><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0111-9.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0111-9.jpg" data-id="708" class="wp-image-708"/></a><figcaption class="blocks-gallery-item__caption"><p>Caption</p></figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/1487352383_mu-shsjvgty-4-1.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/1487352383_mu-shsjvgty-4-1.jpg" data-id="415" class="wp-image-415"/></a><figcaption class="blocks-gallery-item__caption">Alşksdf şlkas dolaş dfasd şad fsa<br>Asf fasd fas<br>A sdfasdf sadf<br> Asdf</figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0003-5.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0003-5.jpg" data-id="701" class="wp-image-701"/></a><figcaption class="blocks-gallery-item__caption">Hello<br>World</figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0005-6.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0005-6.jpg" data-id="728" class="wp-image-728"/></a><figcaption class="blocks-gallery-item__caption">Asdf sadf asdf fşilaks dfişkl asdAsdf asf asdf<br>Asdf</figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0002-1-2.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0002-1-2.jpg" data-id="684" class="wp-image-684"/></a><figcaption class="blocks-gallery-item__caption">Caption yazıyorum gözlerim kapalı asdf asdf asf asf asdf asdf asdf asdf sd Asdf asdf asf f<br>Asf asdf <br>As fasdf<br>Asf fasdf </figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0004-7.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0004-7.jpg" data-id="682" class="wp-image-682"/></a><figcaption class="blocks-gallery-item__caption">Selale selale asdf hklajsdhf lkajsdf lkajshdfl kasjdfhlkajhdf lkajsdf lkasjdfhl aksdjfh lkasjdhf laksjdhf lkasjhdfl ksajdfha lksdfhklasdjh ;Hal;skfjla;skinflint ;lkajsdf ;lkajsdf ;laskdjf </figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/7fe66fb63510d8798a9a5505838cecf2-2.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/7fe66fb63510d8798a9a5505838cecf2-2.jpg" data-id="416" class="wp-image-416"/></a></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/a3022813-7005-47b9-92e6-1004d36e3803.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/a3022813-7005-47b9-92e6-1004d36e3803.jpg" data-id="555" class="wp-image-555"/></a></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/11/img_0111.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/11/img_0111.jpg" data-id="737" class="wp-image-737"/></a></figure></li></ul></figure>
<!-- /wp:gallery -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph -->
```

Reply:
Interestingly I can't repro this on iOS simulators or in my Android device(Huawei p20 lite). I can only repro this with my Pixel 2XL emulator. I also checked web side with dev tools and it is not reproducible there either. Considering `We are currently using this logic as well in gallery.native.js` it seems like we don't have to block for this.